Mathlib/Analysis/Calculus/Gradient/Basic.lean

```diff
@@ -0,0 +1,370 @@
+* the theorems about the congruence of the gradient.
+* the theorems about the gradient of constant function.
+* the theorems about the continuity of a function admitting a gradient.
+-/
+
+open Topology InnerProductSpace Set
+
+noncomputable section
+
+variable {𝕜 F : Type*} [IsROrC 𝕜]
+
+variable [NormedAddCommGroup F] [InnerProductSpace 𝕜 F] [CompleteSpace F]
+
+variable {f : F → 𝕜} {f' x : F}
+
+/-- A function `f` has the gradient `f'` as derivative along the filter `L` if
+  `f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` when `x'` converges along the filter `L`.-/
+def HasGradientAtFilter (f : F → 𝕜) (f' x : F) (L : Filter F) :=
+  HasFDerivAtFilter f (toDual 𝕜 F f') x L
+
+/-- `f` has the gradient `f'` at the point `x` within the subset `s` if
+  `f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x` inside `s`.-/
+def HasGradientWithinAt (f : F → 𝕜) (f' : F) (s : Set F) (x : F):=
+  HasGradientAtFilter f f' x (𝓝[s] x)
+
+/-- `f` has the gradient `f'` at the point `x` if
+  `f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x`. -/
+def HasGradientAt (f : F → 𝕜) (f' x : F) :=
+  HasGradientAtFilter f f' x (𝓝 x)
+
+/-- Gradient of `f` at the point `x`, if it exists.  Zero otherwise.
+If the derivative exists (i.e., `∃ f', HasGradientAt f f' x`), then
+`f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x`. -/
+def Gradient (f : F → 𝕜) (x : F) : F := (toDual 𝕜 F).symm (fderiv 𝕜 f x)
+
+local notation "∇" => Gradient
+
+variable {s : Set F} {L : Filter F}
+
+theorem HasGradientWithinAt_iff_HasFDerivWithinAt {s : Set F} :
+  HasGradientWithinAt f f' s x ↔ HasFDerivWithinAt f (toDual 𝕜 F f') s x := by
+  constructor <;> (intro h; exact h)
+
+theorem HasGradientWithinAt.hasFDerivWithinAt {s : Set F} (h : HasGradientWithinAt f f' s x) :
+  HasFDerivWithinAt f (toDual 𝕜 F f') s x := HasGradientWithinAt_iff_HasFDerivWithinAt.mp h
+
+theorem HasFDerivWithinAt.hasGradientWithinAt {s : Set F}
+  (h : HasFDerivWithinAt f (toDual 𝕜 F f') s x) :
+  HasGradientWithinAt f f' s x := HasGradientWithinAt_iff_HasFDerivWithinAt.mpr h
+
+theorem HasGradientAt_iff_HasFDerivAt :
+  HasGradientAt f f' x ↔ HasFDerivAt f (toDual 𝕜 F f') x := by
+  constructor <;> (intro h; exact h)
+
+theorem HasGradientAt.hasFDerivAt (h : HasGradientAt f f' x) :
+  HasFDerivAt f (toDual 𝕜 F f') x := HasGradientAt_iff_HasFDerivAt.mp h
+
+theorem HasFDerivAt.hasGradientAt (h : HasFDerivAt f (toDual 𝕜 F f') x) :
+  HasGradientAt f f' x := HasGradientAt_iff_HasFDerivAt.mpr h
+
+theorem Gradient_zero_of_not_differentiableAt (h : ¬DifferentiableAt 𝕜 f x) : ∇ f x = 0 := by
+  have : (toDual 𝕜 F).symm 0 = 0 := by simp only [map_zero]
+  rw [← fderiv_zero_of_not_differentiableAt h] at this
+  exact this
+
+local notation "⟪" x ", " y "⟫" => @inner 𝕜 _ _ x y
+
+theorem toDual_eq_inner (y x : F) : ((toDual 𝕜 F) y) x = ⟪y, x⟫ := rfl
+
+theorem toDual_of_toDual_frechet (frechet : F →L[𝕜] 𝕜) :
+  frechet = (toDual 𝕜 F) ((toDual 𝕜 F).symm frechet) := by
```

Comment:
the second and next lines in theorem statements should be indented by 4 spaces. But here you should just remove this theorem: it is just a particular case of `LinearIsometryEquiv.apply_symm_apply` (as you can find by trying `exact?` to see if your theorem can be proved directly).